Using LoRa.cad(...) in a loop makes the code panick?

Hi I saw the cad and tried to experiment with some simple collision detection. It works fine when you just check LoRa.cad, but if i try to repeat it, it crashes.

I found that putting the cad detection in a loop:
```rust
loop {
    info!("Getting CAD ...");
    let mdltn_params = {
        match lora.create_modulation_params(
            SpreadingFactor::_10,
            Bandwidth::_250KHz,
            CodingRate::_4_8,
            LORA_FREQUENCY_IN_HZ,
        ) {
            Ok(mp) => mp,
            Err(err) => {
                info!("Radio error = {}", err);
                continue;
            }
        }
    };

    match lora.prepare_for_cad(&mdltn_params).await {
        Ok(()) => {}
        Err(err) => {
            info!("Radio error = {}", err);
            continue;
        }
    };

    match lora.cad(&mdltn_params).await {
        Ok(cad_activity_detected) => {
            if cad_activity_detected {
                info!("cad successful with activity detected")
            } else {
                info!("cad successful without activity detected")
            }
            Timer::after_secs(5).await;
        }
        Err(err) => info!("cad unsuccessful = {}", err),
    }
}
```

makes the code panic:
```
5.044665 [ERROR] panicked at /home/jakob/repos/pr-lora-rs/lora-phy/src/lib.rs:422:26:
internal error: entered unreachable code (panic_probe panic-probe-1.0.0/src/lib.rs:104)
Firmware exited unexpectedly: Multiple
 WARN probe_rs_debug::debug_info: No reliable return address is available, so we cannot determine the program counter to unwind the previous frame.
Core 0
    Frame 0: HardFault_ @ 0x08009828
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/cortex-m-rt-0.7.5/src/lib.rs:1103:1
    Frame 1: HardFault <Cause: Escalated UsageFault <Cause: Undefined instruction>> @ 0x080049a2

    Frame 2: __udf @ 0x080049a2 inline
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/cortex-m-0.7.7/src/../asm/inline.rs:181:5
    Frame 3: udf @ 0x00000000080049a2 inline
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/cortex-m-0.7.7/src/call_asm.rs:11:43
    Frame 4: hard_fault @ 0x00000000080049a2 inline
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/panic-probe-1.0.0/src/lib.rs:86:5
    Frame 5: panic @ 0x000000000800499a
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/panic-probe-1.0.0/src/lib.rs:54:9
Error: Multiple
```
I have made a reproducible example in my [fork](https://github.com/jakobolason/lora-rs/blob/cad-crashing/examples/stm32wle5cx/src/bin/lora_cad.rs).

I made some advanced debugging (printing everything) in `get_irq_state` and `process_irq_state` to understand what the state changes were, and i can see that the match in line 410 from lib.rs matches the return value:
```rust
match self
    .radio_kind
    .process_irq_event(self.radio_mode, Some(&mut cad_activity_detected), true)
    .await
{
    Ok(Some(IrqState::Done)) => Ok(cad_activity_detected),
    Err(err) => {
        self.radio_kind.ensure_ready(self.radio_mode).await?;
        self.radio_kind.set_standby().await?;
        self.radio_mode = RadioMode::Standby;
        Err(err)
    }
    Ok(_) => unreachable!(),
}
```
But `process_irq_event` might return a `IrqState` which is none, if the cad is not done, from  `get_irq_state(.)`:940 (`sx126/mod.rs`)
```rust
RadioMode::ChannelActivityDetection => {
    if IrqMask::CADDone.is_set(irq_flags) {
        if let Some(detected) = cad_activity_detected {
            *detected = IrqMask::CADActivityDetected.is_set(irq_flags);
        }
        return Ok(Some(IrqState::Done));
    } else {
        trace!("[ERROR] Irq mask CADDone is not set!");
    }
}
```
Which is what my beautiful print statements show:
```
5.044511 [TRACE] [ERROR] Irq mask CADDone is not set! (lora_phy src/sx126x/mod.rs:947)
5.044569 [TRACE] write: [02, ff, ff] (lora_phy lora-phy/src/interface.rs:23)
5.044628 [TRACE] STATE IS NONE (lora_phy src/sx126x/mod.rs:988)
5.044665 [ERROR] panicked at /home/jakob/repos/pr-lora-rs/lora-phy/src/lib.rs:422:26:
```

Now I don't understand what is happening at all, I'm just tracing the state, so I would love to hear if I am using the library wrong?

Yeah, I can confirm, that we hit the `Ok(_) => unreachable!()` branch:
https://github.com/lora-rs/lora-rs/blob/152cb9261ce5a68c8efa4b0884e07223fea8d4f4/lora-phy/src/lib.rs#L410-L422

Log:
```
0.022832 [DEBUG] channel = 868000000 (lora_phy src/sx126x/mod.rs:631)
0.023481 [TRACE] write: [86, 36, 40, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
0.024165 [TRACE] write: [08, 01, 80, 01, 80, 00, 00, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
0.024981 [TRACE] write: [0d, 08, ac, 94] (lora_phy lora-phy/src/interface.rs:23)
0.025668 [TRACE] write: [88, 03, 17, 0a, 00, 00, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
0.026388 [TRACE] write: [c5] (lora_phy lora-phy/src/interface.rs:23)
0.062857 [TRACE] read: addr=[12], len=2, status=a2, buf=[00, 80] (lora_phy lora-phy/src/interface.rs:85)
0.063456 [DEBUG] process_irq satisfied: irq_flags = 0x80 in radio mode ChannelActivityDetection (lora_phy src/sx126x/mod.rs:898)
0.064144 [TRACE] write: [02, ff, ff] (lora_phy lora-phy/src/interface.rs:23)
0.064528 [TRACE] state is Done (lora_phy src/sx126x/mod.rs:987)
0.064823 [INFO ] cad successful without activity detected (lora_cad src/bin/lora_cad.rs:90)
5.065171 [INFO ] Getting CAD ... (lora_cad src/bin/lora_cad.rs:61)
5.065797 [TRACE] write: [80, 00] (lora_phy lora-phy/src/interface.rs:23)
5.066206 [DEBUG] sf = 10, bw = 5, cr = 4 (lora_phy src/sx126x/mod.rs:518)
5.066962 [TRACE] write: [8b, 0a, 05, 04, 00] (lora_phy lora-phy/src/interface.rs:23)
5.068088 [TRACE] read: addr=[1d, 08, 89, 00], len=1, data=[24] (lora_phy lora-phy/src/interface.rs:60)
5.068971 [TRACE] write: [0d, 08, 89, 24] (lora_phy lora-phy/src/interface.rs:23)
5.069421 [DEBUG] channel = 868000000 (lora_phy src/sx126x/mod.rs:631)
5.070094 [TRACE] write: [86, 36, 40, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
5.070803 [TRACE] write: [08, 01, 80, 01, 80, 00, 00, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
5.071643 [TRACE] write: [0d, 08, ac, 94] (lora_phy lora-phy/src/interface.rs:23)
5.072355 [TRACE] write: [88, 03, 17, 0a, 00, 00, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
5.073100 [TRACE] write: [c5] (lora_phy lora-phy/src/interface.rs:23)
5.074627 [TRACE] read: addr=[12], len=2, status=d2, buf=[00, 00] (lora_phy lora-phy/src/interface.rs:85)
5.075226 [DEBUG] process_irq satisfied: irq_flags = 0x0 in radio mode ChannelActivityDetection (lora_phy src/sx126x/mod.rs:898)
5.075622 [TRACE] [ERROR] Irq mask CADDone is not set! (lora_phy src/sx126x/mod.rs:947)
5.076155 [TRACE] write: [02, ff, ff] (lora_phy lora-phy/src/interface.rs:23)
5.076564 [TRACE] STATE IS NONE (lora_phy src/sx126x/mod.rs:988)
5.076814 [ERROR] panicked at /lora-rs/lora-phy/src/lib.rs:422:26:
```

It crashes for me sadly, with this adjusted p2p receive for me at least

[lora_p2p_receive.txt](https://github.com/user-attachments/files/27114884/lora_p2p_receive.txt)

```
0.000009 [INFO ] config done... (lora_p2p_receive src/bin/lora_p2p_receive.rs:44)
0.000092 [TRACE] write: [c0, 00] (lora_phy lora-phy/src/interface.rs:23)
0.003651 [TRACE] write: [80, 00] (lora_phy lora-phy/src/interface.rs:23)
0.003723 [TRACE] write: [96, 01] (lora_phy lora-phy/src/interface.rs:23)
0.003793 [TRACE] write: [8a, 01] (lora_phy lora-phy/src/interface.rs:23)
0.003864 [TRACE] write: [0d, 07, 40, 34, 44] (lora_phy lora-phy/src/interface.rs:23)
0.003945 [TRACE] write: [8f, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
0.004060 [TRACE] read: addr=[1d, 02, 9f, 00], len=9, data=[00, 19, f9, 19, 4f, 70, 8d, a9, 71] (lora_phy lora-phy/src/interface.rs:60)
0.004216 [TRACE] write_buf: [0d, 02, 9f] -> [01, 08, ac, 19, 4f, 70, 8d, a9, 71] (lora_phy lora-phy/src/interface.rs:41)
0.004377 [TRACE] read: addr=[1d, 02, 9f, 00], len=9, data=[01, 08, ac, 19, 4f, 70, 8d, a9, 71] (lora_phy lora-phy/src/interface.rs:60)
0.004533 [TRACE] write_buf: [0d, 02, 9f] -> [02, 08, ac, 08, 89, 70, 8d, a9, 71] (lora_phy lora-phy/src/interface.rs:41)
0.004650 [TRACE] write: [95, 01, 00, 01, 01] (lora_phy lora-phy/src/interface.rs:23)
0.004727 [TRACE] write: [8e, 03, 04] (lora_phy lora-phy/src/interface.rs:23)
0.004815 [TRACE] write: [08, ff, ff, ff, ff, 00, 00, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
0.004884 [INFO ] Getting CAD ... (lora_p2p_receive src/bin/lora_p2p_receive.rs:79)
0.004942 [TRACE] write: [98, d7, db] (lora_phy lora-phy/src/interface.rs:23)
0.006511 [DEBUG] sf = 10, bw = 5, cr = 4 (lora_phy src/sx126x/mod.rs:518)
0.006597 [TRACE] write: [8b, 0a, 05, 04, 00] (lora_phy lora-phy/src/interface.rs:23)
0.006721 [TRACE] read: addr=[1d, 08, 89, 00], len=1, data=[24] (lora_phy lora-phy/src/interface.rs:60)
0.006824 [TRACE] write: [0d, 08, 89, 24] (lora_phy lora-phy/src/interface.rs:23)
0.006880 [DEBUG] channel = 868000000 (lora_phy src/sx126x/mod.rs:631)
0.006953 [TRACE] write: [86, 36, 40, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
0.007065 [TRACE] write: [08, 01, 80, 01, 80, 00, 00, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
0.007152 [TRACE] write: [0d, 08, ac, 94] (lora_phy lora-phy/src/interface.rs:23)
0.007240 [TRACE] write: [88, 03, 17, 0a, 00, 00, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
0.007322 [TRACE] write: [c5] (lora_phy lora-phy/src/interface.rs:23)
0.042424 [TRACE] read: addr=[12], len=2, status=a2, buf=[00, 80] (lora_phy lora-phy/src/interface.rs:85)
0.042513 [DEBUG] process_irq satisfied: irq_flags = 0x80 in radio mode ChannelActivityDetection (lora_phy src/sx126x/mod.rs:898)
0.042595 [TRACE] write: [02, ff, ff] (lora_phy lora-phy/src/interface.rs:23)
0.042652 [INFO ] cad successful without activity detected (lora_p2p_receive src/bin/lora_p2p_receive.rs:94)
5.042693 [INFO ] Getting CAD ... (lora_p2p_receive src/bin/lora_p2p_receive.rs:79)
5.042755 [TRACE] write: [80, 00] (lora_phy lora-phy/src/interface.rs:23)
5.042811 [DEBUG] sf = 10, bw = 5, cr = 4 (lora_phy src/sx126x/mod.rs:518)
5.042903 [TRACE] write: [8b, 0a, 05, 04, 00] (lora_phy lora-phy/src/interface.rs:23)
5.043026 [TRACE] read: addr=[1d, 08, 89, 00], len=1, data=[24] (lora_phy lora-phy/src/interface.rs:60)
5.043135 [TRACE] write: [0d, 08, 89, 24] (lora_phy lora-phy/src/interface.rs:23)
5.043198 [DEBUG] channel = 868000000 (lora_phy src/sx126x/mod.rs:631)
5.043276 [TRACE] write: [86, 36, 40, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
5.043387 [TRACE] write: [08, 01, 80, 01, 80, 00, 00, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
5.043481 [TRACE] write: [0d, 08, ac, 94] (lora_phy lora-phy/src/interface.rs:23)
5.043573 [TRACE] write: [88, 03, 17, 0a, 00, 00, 00, 00] (lora_phy lora-phy/src/interface.rs:23)
5.043661 [TRACE] write: [c5] (lora_phy lora-phy/src/interface.rs:23)
5.043961 [TRACE] read: addr=[12], len=2, status=d2, buf=[00, 00] (lora_phy lora-phy/src/interface.rs:85)
5.044051 [DEBUG] process_irq satisfied: irq_flags = 0x0 in radio mode ChannelActivityDetection (lora_phy src/sx126x/mod.rs:898)
5.044132 [TRACE] write: [02, ff, ff] (lora_phy lora-phy/src/interface.rs:23)
5.044192 [ERROR] panicked at /home/jakob/lora-rs/lora-phy/src/lib.rs:422:26:
internal error: entered unreachable code (panic_probe panic-probe-0.3.2/src/lib.rs:104)
Firmware exited unexpectedly: Multiple
Core 0
    Frame 0: HardFault_ @ 0x08009544
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/cortex-m-rt-0.7.5/src/lib.rs:1103:1
    Frame 1: HardFault <Cause: Escalated UsageFault <Cause: Undefined instruction>> @ 0x08004a0a

    Frame 2: __udf @ 0x08004a0a inline
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/cortex-m-0.7.7/src/../asm/inline.rs:181:5
    Frame 3: udf @ 0x0000000008004a0a inline
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/cortex-m-0.7.7/src/call_asm.rs:11:43
    Frame 4: hard_fault @ 0x0000000008004a0a inline
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/panic-probe-0.3.2/src/lib.rs:86:5
    Frame 5: panic @ 0x0000000008004a02
       /home/jakob/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/panic-probe-0.3.2/src/lib.rs:54:9
```

--- a/examples/stm32wl/build.rs
+++ b/examples/stm32wl/build.rs
@@ -82,4 +82,3 @@ fn main() {
     println!("cargo:rustc-link-arg-bins=-Tlink.x");
     println!("cargo:rustc-link-arg-bins=-Tdefmt.x");
 }
-

--- a/examples/stm32wl/src/iv.rs
+++ b/examples/stm32wl/src/iv.rs
@@ -1,3 +1,4 @@
+use cortex_m::peripheral::NVIC;
 use embassy_stm32::interrupt::InterruptExt;
 use embassy_stm32::{interrupt, pac};
 use embassy_sync::blocking_mutex::raw::CriticalSectionRawMutex;
@@ -67,6 +68,8 @@ where
     }
 
     async fn await_irq(&mut self) -> Result<(), RadioError> {
+        // Clear pending interrupts before enabling IRQ
+        NVIC::unpend(pac::Interrupt::SUBGHZ_RADIO);
         unsafe { interrupt::SUBGHZ_RADIO.enable() };
         IRQ_SIGNAL.wait().await;
         Ok(())

--- a/examples/stm32wle5cx/src/iv.rs
+++ b/examples/stm32wle5cx/src/iv.rs
@@ -1,3 +1,4 @@
+use cortex_m::peripheral::NVIC;
 use embassy_stm32::interrupt::InterruptExt;
 use embassy_stm32::{interrupt, pac};
 use embassy_sync::blocking_mutex::raw::CriticalSectionRawMutex;
@@ -5,10 +6,10 @@ use embassy_sync::signal::Signal;
 use embassy_time::Timer;
 use embedded_hal::digital::OutputPin;
 use embedded_hal_async::spi::{ErrorType, Operation, SpiBus, SpiDevice};
+use lora_phy::DelayNs;
 use lora_phy::mod_params::RadioError;
 use lora_phy::mod_params::RadioError::*;
 use lora_phy::mod_traits::InterfaceVariant;
-use lora_phy::DelayNs;
 
 /// Interrupt handler.
 pub struct InterruptHandler {}
@@ -67,6 +68,8 @@ where
     }
 
     async fn await_irq(&mut self) -> Result<(), RadioError> {
+        // Clear pending interrupts before enabling IRQ
+        NVIC::unpend(pac::Interrupt::SUBGHZ_RADIO);
         unsafe { interrupt::SUBGHZ_RADIO.enable() };
         IRQ_SIGNAL.wait().await;
         Ok(())

